hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/listener/ClientClusterViewListenerService.java

```diff
@@ -0,0 +1,136 @@
+        this.connectionManager = client.getConnectionManager();
+        partitionService = (ClientPartitionServiceImpl) client.getClientPartitionService();
+        clusterService = (ClientClusterServiceImpl) client.getClientClusterService();
+    }
+
+    public void start() {
+        connectionManager.addConnectionListener(this);
+    }
+
+    private final class ClusterViewListenerHandler extends ClientAddClusterViewListenerCodec.AbstractEventHandler
+            implements EventHandler<ClientMessage> {
+
+        private final Connection connection;
+
+        private ClusterViewListenerHandler(Connection connection) {
+            this.connection = connection;
+        }
+
+        @Override
+        public void beforeListenerRegister(Connection connection) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Register attempt of ClusterViewListenerHandler to " + connection);
+            }
+        }
+
+        @Override
+        public void onListenerRegister(Connection connection) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Registered ClusterViewListenerHandler to " + connection);
+            }
+        }
+
+        @Override
+        public void handleMembersViewEvent(int memberListVersion, Collection<MemberInfo> memberInfos) {
+            clusterService.handleMembersViewEvent(memberListVersion, memberInfos);
+        }
+
+        @Override
+        public void handlePartitionsViewEvent(int version, Collection<Map.Entry<Address, List<Integer>>> partitions) {
+            partitionService.handlePartitionsViewEvent(connection, partitions, version);
+        }
+    }
+
+    @Override
+    public void connectionAdded(Connection connection) {
+        if (listenerAddedConnection.compareAndSet(null, connection)) {
+            register(connection);
+        }
+    }
+
+    @Override
+    public void connectionRemoved(Connection connection) {
+        if (listenerAddedConnection.compareAndSet(connection, null)) {
+            ((ClientListenerServiceImpl) listenerService).removeEventHandler(lastCorrelationId);
+            Connection newConnection = connectionManager.getRandomConnection();
+            if (newConnection != null) {
+                if (listenerAddedConnection.compareAndSet(null, newConnection)) {
+                    register(newConnection);
+                }
+            }
+        }
+    }
+
+    private void register(Connection connection) {
+        ClientMessage clientMessage = ClientAddClusterViewListenerCodec.encodeRequest();
+        ClientInvocation invocation = new ClientInvocation(client, clientMessage, null, connection);
+        ClusterViewListenerHandler handler = new ClusterViewListenerHandler(connection);
+        invocation.setEventHandler(handler);
+        handler.beforeListenerRegister(connection);
+        invocation.invokeUrgent().thenAcceptAsync(message -> {
```

Comment:
why not call `thenAccept` since the executor is `CALLER_RUNS `?

Reply:
Because the affect of `thenAccept` will be different from CALLER_RUNS in near future. 
https://github.com/hazelcast/hazelcast/pull/16154